### PR TITLE
describe the user

### PR DESCRIPTION
**Pull request by heroicNeZha**
#### What type of PR is this?
- [ ] bug
- [x] feature
- [ ] enhancement

#### What does this PR do?
$ desc user endy
+---------+------------------------------------+
| Account | Roles in spaces                    |
+---------+------------------------------------+
| "endy"  | " USER in nba, USER in my_space_1" |
+---------+------------------------------------+
Got 1 rows (time spent 1319/6278 us)

#### Which issue(s)/PR(s) this PR relates to?
how to describe the user? #2574
https://github.com/vesoft-inc/nebula/issues/2574
  
#### Checklist：
- [ ] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/parser/scanner.lex [hunk 1]
```diff
@@ -105,6 +104,7 @@ IP_OCTET                    ([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])
 "NO"                        { return TokenType::KW_NO; }
 "OVERWRITE"                 { return TokenType::KW_OVERWRITE; }
 "SHOW"                      { return TokenType::KW_SHOW; }
+"DESCRIBE"                  { return TokenType::KW_DESCRIBE; }
```
Comment (yixinglu) [inline src/parser/scanner.lex:107]: Why change the location of the keyword?

File: src/graph/validator/ACLValidator.cpp [hunk 2]
```diff
@@ -131,6 +131,21 @@ Status RevokeRoleValidator::toPlan() {
                                        sentence->getAclItemClause()->getRoleType());
 }
 
+// describe user
+Status DescribeUserValidator::validateImpl() {
+  auto sentence = static_cast<DescribeUserSentence *>(sentence_);
+  if (sentence->account()->size() > kUsernameMaxLength) {
```
Comment (yixinglu) [inline src/graph/validator/ACLValidator.cpp:137]: You should check whether the current user has the permission to describe the `account`.

File: src/graph/service/PermissionManager.cpp [hunk 3]
```diff
@@ -113,6 +113,26 @@ Status PermissionManager::canWriteUser(ClientSession *session) {
...
+  }
+  if (session->isGod()) {
+    return Status::OK();
+  }
+
+  if (session->user() == targetUser) {
+    return Status::OK();
+  }
+
+  return Status::PermissionError("No permission to read user.");
```
Comment (yixinglu) [inline src/graph/service/PermissionManager.cpp:133]: `No permission to read 'which' user!`

**Review by yixinglu — COMMENTED** [review]
Could you add some TCK test cases for `DESCRIBE USER` ? 
- root describe guest and root
- admin describe other admin
- guest describe root or admin

**Review by CPWstatic — APPROVED** [review]

**Comment by heroicNeZha** [comment]
> Could you add some TCK test cases for `DESCRIBE USER` ?
> 
> * root describe guest and root
> * admin describe other admin
> * guest describe root or admin

done

File: tests/tck/features/user/User.feature [hunk 4]
```diff
@@ -280,3 +280,85 @@ Feature: User & privilege Test
...
+    Then the result should be, in any order, with relax comparison:
+      | Account | Roles in spaces     |
+      | "root"  | " GOD in ALL_SPACE" |
+    When executing query:
+      """
+      DESC USER user1
+      """
+    Then the result should be, in any order, with relax comparison:
+      | Account | Roles in spaces              |
+      | "user1" | " ADMIN in user_tmp_space_4" |
```
Comment (yixinglu) [inline tests/tck/features/user/User.feature:333]: Maybe following table format is better:

 | Role | Space |
|-----|----|
 | ADMIN | tmp_space_4 |
| USER | tmp_space_2 |
| GUEST | space3 |

They can use pipe to further filter what they want. WDYT? For example:

```sql
desc user user1 | YIELD $-.role where $-.space == "space3" 
```

In addition, do we have `DESC USERS` commands to show all users for root? if exists, its output table maybe:

 | USER | Role | Space |
|-----|-----|----|
| user1 | ADMIN | tmp_space_4 |
| user1 | USER | tmp_space_2 |
| user2 | GUEST | space3 |

File: tests/conftest.py [hunk 5]
```diff
@@ -109,6 +109,17 @@ def conn_pool_to_second_graph_service(pytestconfig):
     yield pool
     pool.close()
 
+
+@pytest.fixture(scope="session")
+def conn_pool_to_third_graph_service(pytestconfig):
```
Comment (yixinglu) [inline tests/conftest.py:114]: Is the third graphd service connection pool necessary to test your cases? Why not to use the existing conn pool?

Comment (heroicNeZha) [reply tests/conftest.py:114]: the existing conn pool is root account，my cases need guest account

**Review by yixinglu — DISMISSED** [review]

**Review by yixinglu — DISMISSED** [review]

File: src/meta/processors/user/AuthenticationProcessor.cpp [hunk 6]
```diff
@@ -252,6 +252,47 @@ void ListUsersProcessor::process(const cpp2::ListUsersReq& req) {
   onFinished();
 }
 
+void DescribeUserProcessor::process(const cpp2::DescribeUserReq& req) {
+  UNUSED(req);
```
Comment (Shylock-Hg) [inline src/meta/processors/user/AuthenticationProcessor.cpp:256]: Remove this line.

File: src/interface/meta.thrift [hunk 7]
```diff
@@ -1201,6 +1218,7 @@ service MetaService {
     ExecResp grantRole(1: GrantRoleReq req);
     ExecResp revokeRole(1: RevokeRoleReq req);
     ListUsersResp listUsers(1: ListUsersReq req);
+    DescribeUserResp describeUser(1: DescribeUserReq req);
```
Comment (Shylock-Hg) [inline src/interface/meta.thrift:1221]: This is duplicate with RPC getUserRoles

Comment (yixinglu) [reply src/interface/meta.thrift:1221]: @heroicNeZha should make sure whether these two interfaces do the same thing. If so, plz don't define duplicate interfaces.

Comment (heroicNeZha) [reply src/interface/meta.thrift:1221]: sure, I will remove it

File: src/graph/service/PermissionManager.cpp [hunk 8]
```diff
@@ -113,6 +113,26 @@ Status PermissionManager::canWriteUser(ClientSession *session) {
...
 }
 
+// static
+Status PermissionManager::canReadUser(ClientSession *session, const std::string &targetUser) {
+  if (!FLAGS_enable_authorize) {
+    return Status::OK();
+  }
+  // Cloud auth user cannot create user
+  if (FLAGS_auth_type == "cloud") {
+    return Status::PermissionError("Cloud authenticate user can't write user.");
```
Comment (yixinglu) [inline src/graph/service/PermissionManager.cpp:123]: can't read user ?

File: src/graph/service/PermissionManager.cpp [hunk 9]
```diff
@@ -113,6 +113,26 @@ Status PermissionManager::canWriteUser(ClientSession *session) {
   }
 }
 
+// static
+Status PermissionManager::canReadUser(ClientSession *session, const std::string &targetUser) {
+  if (!FLAGS_enable_authorize) {
+    return Status::OK();
+  }
+  // Cloud auth user cannot create user
```
Comment (yixinglu) [inline src/graph/service/PermissionManager.cpp:121]: cleanup the comment or let it be right.

File: src/parser/parser.yy [hunk 10]
```diff
@@ -2714,6 +2723,7 @@ traverse_sentence
     | delete_edge_sentence { $$ = $1; }
     | show_queries_sentence { $$ = $1; }
     | kill_query_sentence { $$ = $1; }
+    | describe_user_sentence { $$ = $1; }
```
Comment (Shylock-Hg) [inline src/parser/parser.yy:2726]: The traverse_sentence means supports input/output. Don't it here.

Comment (heroicNeZha) [reply src/parser/parser.yy:2726]: The desc_user_sentcence supports output pipe.

Comment (Shylock-Hg) [reply src/parser/parser.yy:2726]: Why need this? Normally, admin sentence don't support it.

Comment (heroicNeZha) [reply src/parser/parser.yy:2726]: To Support `desc user user1 | YIELD $-.role where $-.space == "space3" `

Comment (Shylock-Hg) [reply src/parser/parser.yy:2726]: Does this pass in discussion meeting?

Comment (heroicNeZha) [reply src/parser/parser.yy:2726]: @yixinglu From Yee‘s conversation

Comment (heroicNeZha) [reply src/parser/parser.yy:2726]: @yixinglu From Yee‘s conversation

**Review by Shylock-Hg — APPROVED** [review]
LGMT except allowing input/output. How do you think about it? @CPWstatic

File: tests/tck/features/user/User.feature [hunk 11]
```diff
@@ -280,3 +280,89 @@ Feature: User & privilege Test
...
+    When executing query:
+      """
+      DESC USER user1
+      """
+    Then the result should be, in any order, with relax comparison:
+      | role    | space              |
+      | "ADMIN" | "user_tmp_space_4" |
+    When executing query:
+      """
+      DESC USER user1 | YIELD $-.space as sp
```
Comment (yixinglu) [inline tests/tck/features/user/User.feature:336]: Maybe more tests about: `SHOW USERS | DESC USER $-.Account`

**Review by yixinglu — APPROVED** [review]

**Pull request merged**